src/ast/SemanticChecksVisitor.java

```diff
@@ -10,10 +10,17 @@
     private Map<String, List<STSymbol>> instanceTemplates;
     private boolean isLegalForest;
     private boolean isLegalST;
+    private boolean visitResult; //should be set to true at builder but any subsequent sets are to false
```

Comment:
Didn't understand the comment.

Reply:
We start with the assumption that the program is legal.
Once we discover otherwise, we shouldn't let any subsequent checks to reset the visit result to true.
So it is set true at the start but than any other assignment to it must be do to an error so it should remain false forever